### ISSUE TITLE
Cannot create new experiment with previously-used Experiment Site ID

### ISSUE THREAD

**Issue by amurphy-cl** [CarnegieLearningWeb/educational-experiment-service#95]
See screenshot. Was working to QA a concept builder experiment, and I had made a similar one a couple months ago and enrollment was completed. When I tried to create the new experiment, I saw an error that the Experiment Site was already in use (despite the old experiment no longer running). We need a better way of handling the ability to create a new experiment using the same Experiment Site ID as a prior experiment. 
<img width="899" alt="Screen Shot 2020-12-17 at 10 04 26 AM" src="https://user-images.githubusercontent.com/65311369/102527337-f10ffd00-406a-11eb-9bf1-22cd7b5fce0f.png">

**Comment by nirmalpatel** [comment]
Our current solution path is to ensure that there is no experiment point + ID clash in the current "Preview" and "Enrolling" experiments. So we can create an experiment with point + ID, mark it as enrollment complete, create another experiment with the same point + ID, mark that one as complete, and create one more, and so on. When we switch back an experiment from Enroll Complete to Enroll, we again ensure that there are no preview/enrolling experiments having same point + ID as the given experiment.

This will play into how we match metrics data with the experiment assignment data exported form UpGrade, because metric data do not have experiment IDs and they simply have point + ID information. This means that if there have been more than one experiments on a point + ID, we need to look at the experiment start/end/restart times and use that time information to appropriately subset the metric data for that point + ID to get the experiment related metrics.

**Comment by nirmalpatel** [comment]
**Steve:** I don’t think you need to get too smart about the metrics data. If this is a workspace experiment, then (most likely) participants either did the first experiment or the second but not both - so participant inclusion in the experiment will already partition the data. Experiments with general features (like something about the hint window) might need to look at data for a time period (e.g. number of hints during the experiment). I don’t think we can currently do that. But I think that’s a change to the metric specification, not something specific to running multiple experiments with the same point/id

**Comment by nirmalpatel** [comment]
@amurphy-cl @SritterCL, JD and I thought more about this and this looks like a very big change in the system. There are several modules in the backend that are using the globally unique point + ID combination. Two main places are random assignment and markExperimentPoint functionality.

It seems to me that this change will affect the semantics of the markExperimentPoint function. Right now, we have a table which tracks which experiment points and IDs a student has seen, and if we run a two experiments with same point and IDs, the second experiment will exclude the user from the experiment since they have seen the point + ID already. This wouldn't be a desired behavior, unless we want to ensure that the second experiment should only enroll students who haven't seen the first experiment. If we want all of the students to enroll into the second experiment, we can do a workaround for that. Every time an with an experiment becomes active and it has conflicting point + IDs with any inactive experiments, we remove all students' markExperimentPoint entries with the given experiment point + ID combination.

Please let me know what you guys think.

Thanks!

**Comment by jerith666** [comment]
But what if a completed experiment has post-experiment rule "continue"?  Doesn't that require us to keep around records of that experiment's assignments indefinitely?

**Comment by SritterCL** [comment]
I think you're right about the implications. But I also think excluding students from the second experiment would be OK (at least under the cases I can think of). If there are two experiments on, for example, area of a triangle, then students who were in the first experiment would be excluded from the second experiment (assuming we use group cohesion) because the second experiment is only supposed to go to students who haven't already seen area of a triangle.

But we don't really have enough experience to understand the general needs for managing experiments over time. For now, maybe we should just allow people to create a new experiment for the same point+ID (assuming the prior experiment is inactive) but give a warning.

**Comment by SritterCL** [comment]
> But what if a completed experiment has post-experiment rule "continue"? Doesn't that require us to keep around records of that experiment's assignments indefinitely?

Yeah, I was just going to bring that up as well. What counts as "inactive" is tricky. "Archived" is safe. Even enrollment complete without continue might be problematic, since it could reference a particular condition.

**Comment by SritterCL** [comment]
Once we define "archived" (issue 96), we should allow new experiments with the same point and id of an experiment in archived status.

**Comment by nirmalpatel** [comment]
@SritterCL @jerith666 @amurphy-cl So right now, you can re-use a point+ID in an existing experiment after you delete that experiment from the system. If you cancel an experiment, the point+ID remains locked. Is that the desired behavior for the Canceled status? If we want to unlock a point+ID after the related experiment is Canceled, that behavior would be nearly identical to the Archived status, I guess? 

Maybe we can just have a single Stop status that stops the experiment and frees up the point+ID for a new experiment. And once and experiment is stopped, its status cannot be changed and its condition assignments would not be respected. We would still keep the records of students who marked the point+ID in that experiment so that they don't get reenrolled into a new experiment at that point+ID.

Let me know what you guys think.

**Comment by SritterCL** [comment]
Deleting an experiment from the server should mean that the UpGrade server doesn't even know the experiment ever existed. So that's definitely an option for freeing up the point+id. (As a separate issue, I'm not sure *how* to delete an experiment - is that in the UI somewhere?).

The idea of the "archived" status is that the experiment data (and participant data) stays on the UpGrade server, but the experiment is finished and cannot be restarted (this is different from "enrollment complete" - you can transition from EC to "enrolling" if you want to restart the experiment). The UpGrade server shouldn't provide any information about the experiment in response to getAllExperimentConditions.

Cancelled is a special status, with the semantics that a cancelled experiment is one that was started (or, at least, intended to be) but something went wrong. If an experiment has cancelled status, the experiment parameters are ignored - all participants get "default" as their condition, regardless of whether they had previously been assigned to a condition. It's an escape hatch.

So I think the user behavior for both statuses is the same, but they mean different things.

The "stop" status that you're proposing might be useful, but I'm not sure I want to add a new status just to deal with this issue. Part of the semantics of "archived" is that client developers are free to change the code to remove support for the experimental conditions. That seems to be the signal you want to send, if you're going to start a new experiment with the same point+id. So forcing the experiment to go into archived status to do that seems right to me.

For now, though, deleting the experiment seems like a fine way to deal with the issue. "Deleted" isn't really a status. I imagine that archived experiments will eventually get deleted, though when that "clean up" gets done is up to the UpGrade server administrator, I guess.

**Comment by nirmalpatel** [comment]
You can delete any experiment from the UI by pressing the Delete button inside the experiment page:

![image](https://user-images.githubusercontent.com/830400/104034058-639d7400-51f6-11eb-9088-d77ec3a4943f.png)

What you say about Archived makes sense to me. I'm still trying to identify what is the difference between Canceled and Archived except for the label in the UI. If we cancel a running experiment and an already enrolled student calls getAllExperimentConditions, they will not get anything for the canceled experiment's point + ID, right?

**Comment by SritterCL** [comment]
Not sure how I missed that button in the UI:)

I guess the difference between cancelled and archived is that a cancelled experiment is one that ended abnormally. Archived experiments are ones that are over and ended normally.

Both types should maintain data on the server. If you archive an experiment but don't delete it, you can later go back to the experiment and look at the data or the conditions (and for cancelled experiments, you might want that data to diagnose what went wrong).

I updated the state diagrams in the spec - see if it makes sense to you there.
